codec-stomp/src/main/java/io/netty/handler/codec/stomp/StompSubframeEncoder.java

```diff
@@ -19,63 +19,91 @@
 import io.netty.buffer.ByteBufUtil;
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.handler.codec.MessageToMessageEncoder;
-import io.netty.util.CharsetUtil;
 
 import java.util.List;
 import java.util.Map.Entry;
 
+import static io.netty.handler.codec.stomp.StompConstants.*;
+
 /**
  * Encodes a {@link StompFrame} or a {@link StompSubframe} into a {@link ByteBuf}.
  */
 public class StompSubframeEncoder extends MessageToMessageEncoder<StompSubframe> {
 
     @Override
-    protected void encode(ChannelHandlerContext ctx, StompSubframe msg, List<Object> out) throws Exception {
+    protected void encode(ChannelHandlerContext ctx, StompSubframe msg, List<Object> out) {
         if (msg instanceof StompFrame) {
-            StompFrame frame = (StompFrame) msg;
-            ByteBuf frameBuf = encodeFrame(frame, ctx);
-            if (frame.content().isReadable()) {
-                out.add(frameBuf);
-                ByteBuf contentBuf = encodeContent(frame, ctx);
-                out.add(contentBuf);
-            } else {
-                frameBuf.writeByte(StompConstants.NUL);
-                out.add(frameBuf);
-            }
+            StompFrame stompFrame = (StompFrame) msg;
+            ByteBuf buf = encodeFullFrame(stompFrame, ctx);
+
+            out.add(converter().convert(stompFrame, buf));
         } else if (msg instanceof StompHeadersSubframe) {
-            StompHeadersSubframe frame = (StompHeadersSubframe) msg;
-            ByteBuf buf = encodeFrame(frame, ctx);
-            out.add(buf);
+            StompHeadersSubframe stompHeadersSubframe = (StompHeadersSubframe) msg;
+            ByteBuf buf = ctx.alloc().buffer(headersSubFrameSize(stompHeadersSubframe));
+            encodeHeaders(stompHeadersSubframe, buf);
+
+            out.add(converter().convert(stompHeadersSubframe, buf));
         } else if (msg instanceof StompContentSubframe) {
             StompContentSubframe stompContentSubframe = (StompContentSubframe) msg;
             ByteBuf buf = encodeContent(stompContentSubframe, ctx);
-            out.add(buf);
+
+            out.add(converter().convert(stompContentSubframe, buf));
         }
     }
 
-    private static ByteBuf encodeContent(StompContentSubframe content, ChannelHandlerContext ctx) {
-        if (content instanceof LastStompContentSubframe) {
-            ByteBuf buf = ctx.alloc().buffer(content.content().readableBytes() + 1);
-            buf.writeBytes(content.content());
-            buf.writeByte(StompConstants.NUL);
-            return buf;
-        } else {
-            return content.content().retain();
+    protected StompEncodedSubframeConverter<?> converter() {
+        return StompEncodedSubframeConverter.NOOP;
+    }
+
+    /**
+     * Returns a heuristic size for headers (32 bytes per header line) + (2 bytes for colon and eol)
+     * + (additional command buffer).
+     */
+    protected int headersSubFrameSize(StompHeadersSubframe headersSubframe) {
```

Comment:
`private static`

Reply:
I thought to leave it open so that somebody could override it since this is just a heuristic.